kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/checks/utils/Scheduler.kt

```diff
@@ -0,0 +1,90 @@
+package com.kotlindiscord.kord.extensions.checks.utils
+
+import kotlinx.coroutines.*
+import mu.KotlinLogging
+import kotlin.coroutines.EmptyCoroutineContext
+
+private val logger = KotlinLogging.logger {  }
+
+/**
+ * Class in charge of providing scheduling functions.
+ */
+class Scheduler {
+    private val jobMap: MutableMap<Int, Job> = mutableMapOf()
+
+    private val scope = CoroutineScope(EmptyCoroutineContext)
+    private val finishTask = CancellationException()
+
+    /**
+     * Schedule a new task.
+     *
+     * @param id The ID of the task.
+     * @param delay How long in milliseconds to wait before executing the callback.
+     * @param data Arbitrary data to be passed to the callback.
+     * @param callback Function to be executed.
+     *
+     * @return Return true on success, false otherwise.
+     */
+    fun <T> schedule(id: Int, delay: Long, data: T?, callback: (T?) -> Unit): Boolean {
+        logger.debug { "Scheduling task $id" }
+
+        if (id in jobMap) {
+            logger.warn { "Cannot schedule task, ID already in use." }
+            return false
+        }
+
+        val job = scope.launch {
+            delay(delay)
+            callback(data)
```

Comment:
It might be worth making callbacks optional, for jobs where the invoker doesn't need to worry about what happens after.